libmscore/instrument.cpp

```diff
@@ -443,9 +438,13 @@ Channel::~Channel()
 
 void Channel::firePropertyChanged(Channel::Prop prop)
       {
-      QList<ChannelListener *> list(listeners);
+      QList<QPointer<QObject>> list(listeners);
+
+      for (QPointer<QObject> ptr: list) {
+            if (ptr.isNull())
+                  continue;
 
-      for (ChannelListener * l: list) {
+            ChannelListener *l = dynamic_cast<ChannelListener *>(ptr.data());
```

Comment:
And concerning this `dynamic_cast`, either it will be not necessary if the list will have correct type definition (see my first comment) or the result of `dynamic_cast` should be checked on being `nullptr`.

Reply:
The QPointer is already checked for null.  Is it necessary to do it again?

Reply:
Also, I would like to use the correct type too but cannot think of a way to do it using the Qt system.  If you have any ideas, please let me know.

Reply:
Well, the reason for the proposed check is type safety mainly. In case that pointer points to some `QObject` that is not a `ChannelListener` dynamic_cast will return a null pointer. If it can be stated for sure that this situation is impossible then the check is not necessary but this looks not very safe... I will think on what can be done here too.

In fact I believe that accurate usage of `removeListener` should prevent crashes even without this smart pointers system. Currently `addListener`/`removeListener` calls seem to be slightly mismatched, for example in `mixerdetails.cpp`:
https://github.com/musescore/MuseScore/blob/7b0a14d1db0fa056eb01c59f7bde38eec13419f5/mscore/mixerdetails.cpp#L93-L94
https://github.com/musescore/MuseScore/blob/7b0a14d1db0fa056eb01c59f7bde38eec13419f5/mscore/mixerdetails.cpp#L73
Could fixing listeners list management perhaps fix the issue with crashes?

Reply:
I don't think so.  Channels are passed by value often, so there really isn't any one Channel object to delete the listeners from.  By using QPointer you let Qt handle keeping track of the objects.

Reply:
Well, I have another suggestion/remark. As far as I understand from looking at the code each listener is assumed to listen on one channel only, and violating this rule may lead to unpredictable consequences (crashes?). Meanwhile, as you noted, `Channel` objects are often passed by value which probably means they are copied sometimes. Listeners list is also duplicated, and each listener gets bound to two channels at a time.

Such a situation can be prevented by proper definition of copy constructor and copy assignment operators. [The rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(C%2B%2B_programming)) also suggests that we should have such member functions defined (or deleted) for `Channel` class, and violating this rule is not a good practice either way.

If you still need to transfer listeners to another channel by doing assignment in some situations you may define move constructor/move assignment operator as well.

I believe this is anyway good to do, and after doing that chances are good that number of issues related to listeners management will reduce.

Reply:
Having multiple copies of the listener isn't really the problem even though this could cause extra updates.  The real problem is that multiple objects can potentially be connected to a single object as listeners, the listener needs to be removed if either end of the connection is deleted.  You also don't have much control over what order things will be deleted.

Using weak pointers (or QPointers which have similarities) is one way to deal with this - if the 'slot' object is deleted, you can detect this before sending the 'signal' and delete the connection automatically.  If the signal object is deleted, the list of listeners it maintains is deleted automatically.  Since listeners are one directional, this works in a situation where the Channel can't do a lot of the introspection tricks that a QObject can.